Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    public let supportedAuthenticationMethods: [AuthenticationMethod]
+    public let targetAddress: AddressType
+    public let targetPort: UInt16
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    public init(supportedAuthenticationMethods: [AuthenticationMethod], targetAddress: AddressType, targetPort: UInt16) {
```

Comment:
Suppose users pass `usernamePassword` as a valid authentication method: how will they _supply_ that username and password?

Reply:
It's a separate handshake - this is just notifying that if the server wants to it _can_ select that authentication method.

Reply:
Yes, but if it chose it how would users select it?

Stepping back, what I'm concretely asking for is that you have a good picture in your head of how user authentication will work and then we design to that API. We don't need to _implement_ that API, but we need to know where we're going so we don't paint ourselves into a corner.

Reply:
1. Client sends greeting
2. Server receives greeting ->  sends up the channel to the implementor
3. Implementor selects the method -> sends down the channel to _our_ handler
4. Handler packages it and sends it to the client
5. Client receives reply -> sends up the channel to the implementor
6. Client implementor sends a struct `UsernamePasswordAuthentication` or some other type containing the details

Does that make sense?

Reply:
Sends it to the client how?

Reply:
In a message containing 2 bytes - the version (always 5), and the identifying number for the selected authentication mechanism. Is that what you're asking?

Reply:
Sorry, I mean "client implementor".

Reply:
The thing I'm pushing on here is that this handler claims to be `ByteBuffer` in, `ByteBuffer` out, but you're also proposing to send other structured data down the pipeline. That strikes me as overcomplicated.

I don't really see a reason this handler shouldn't just deal with this entirely itself: that is, be configured with the objects that know how to handle auth.

Reply:
Ah, you make a good point. This is an oversight on my part - what about some like in HTTP where we have an enum to represent the parts?

E.g.

```swift
enum ClientSocksPart {
    case selectedAuthMethod(_)
    case data(_)
}
```

Reply:
I think we should consider whether it's worth involving another ChannelHandler in this at all, or whether we should just use a delegate object like we do in SSH.

Reply:
Either work for me, I'll make a delegate

Reply:
Right I've got some for of delegate. I'm going to mark this as resolved and we can start a new thread to make this a little easier to follow (assuming you have comments/question regarding the implementation).